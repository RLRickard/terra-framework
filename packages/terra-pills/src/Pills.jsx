import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import classNamesBind from 'classnames/bind';
import ThemeContext from 'terra-theme-context';
import { injectIntl } from 'react-intl';
import Pill from './Private/_Pill';
import styles from './Private/Pill.module.scss';
import RollUpPill from './Private/_RollupPill';

const cx = classNamesBind.bind(styles);

const propTypes = {
  /**
   * A collection of pills  to render within the Pills container body.
   */
  children: PropTypes.node,
  /**
   * Callback function to remove a pill, return pillKey, metadata
   */
  onRemove: PropTypes.func,
  /**
   * @private
   * The intl object to be injected for translations.
   */
  intl: PropTypes.shape({ formatMessage: PropTypes.func }).isRequired,
};

const Pills = (props) => {
  const {
    children,
    onRemove,
    intl,
    ...customProps
  } = props;

  const theme = React.useContext(ThemeContext);

  const handleOnRemove = (pillKey, metaData) => {
    if (onRemove) {
      onRemove(pillKey, metaData);
    }
  };

  const pillListClassNames = classNames(
    cx([
      'pill-list',
      theme.className,
    ]),
    customProps.className,
  );

  const renderChildren = (items) => React.Children.map(items, (pill) => React.cloneElement(pill, { onRemove: handleOnRemove, isSelectable: true }));

  return (
    <div
      {...customProps}
      className={pillListClassNames}
      tabIndex="-1"
    >
      {children ? renderChildren(children) : []}
      <RollUpPill pillCount={children.length} />
    </div>
  );
};

Pills.Pill = Pill;
Pills.propTypes = propTypes;

export default injectIntl(Pills);
