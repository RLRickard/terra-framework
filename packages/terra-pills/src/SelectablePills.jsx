import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import classNamesBind from 'classnames/bind';
import ThemeContext from 'terra-theme-context';
import Pill from './Private/_Pill';
import styles from './Private/Pill.module.scss';

const cx = classNamesBind.bind(styles);

const propTypes = {
  /**
   * A collection of pills to render within the Selectable Pills Container body.
   */
  children: PropTypes.node,
  /**
   * Callback function to remove a pill, returns pillKey, metadata
   */
  onRemove: PropTypes.func,
  /**
  * Callback function to select a pill, returns pillKey, metadata
  */
  onSelect: PropTypes.func,
};

const SelectablePills = (props) => {
  const {
    children,
    onRemove,
    onSelect,
    ...customProps
  } = props;

  const theme = React.useContext(ThemeContext);
  const selectablePillsRef = useRef();

  const handleOnRemove = (pillKey, metaData) => {
    if (onRemove) {
      onRemove(pillKey, metaData);
    }
  };

  const handleOnSelect = (pillKey, metaData) => {
    if (onSelect) {
      onSelect(pillKey, metaData);
    }
  };

  const pillListClassNames = classNames(
    cx([
      'pill-list',
      theme.className,
    ]),
    customProps.className,
  );

  const renderChildren = (items) => React.Children.map(items, (pill) => React.cloneElement(pill, { onRemove: handleOnRemove, onSelect: handleOnSelect, isSelectable: false }));

  return (
    <div
      {...customProps}
      className={pillListClassNames}
      ref={selectablePillsRef}
      tabIndex="-1"
    >
      {children ? renderChildren(children) : []}
    </div>
  );
};

SelectablePills.Pill = Pill;
SelectablePills.propTypes = propTypes;

export default SelectablePills;
