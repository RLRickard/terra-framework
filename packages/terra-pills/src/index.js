import Pills from './Pills';
import SelectablePills from './SelectablePills';

export default Pills;
export { SelectablePills };
