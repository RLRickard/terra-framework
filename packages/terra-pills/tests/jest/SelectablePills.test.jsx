import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { mountWithIntl } from 'terra-enzyme-intl';
import { SelectablePills } from '../../src/index';

describe('Selectable Pill', () => {
  it('should render a selectable pill', () => {
    const wrapper = mountWithIntl(
      <SelectablePills
        ariaLabel="Example of a Selectable Pill"
        onSelect={() => jest.fn()}
      >
        <SelectablePills.Pill
          label="Selectable Pill"
          id="selectabla-pill"
        />
      </SelectablePills>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a selectable pill with roll up', () => {
    const wrapper = mountWithIntl(
      <SelectablePills
        ariaLabel="Example of a Selectable Pill"
        onSelect={() => jest.fn()}
        isSingleLine
      >
        <SelectablePills.Pill
          label="Selectable Pill"
          id="selectabla-pill"
        />
      </SelectablePills>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
