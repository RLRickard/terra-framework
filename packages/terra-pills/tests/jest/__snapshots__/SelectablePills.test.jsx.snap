// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selectable Pill should render a selectable pill 1`] = `
<InjectIntl(SelectablePills)
  ariaLabel="Example of a Selectable Pill"
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": null,
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
  onSelect={[Function]}
>
  <SelectablePills
    ariaLabel="Example of a Selectable Pill"
    intl={
      Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatHTMLMessage": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatPlural": [Function],
        "formatRelative": [Function],
        "formatTime": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralFormat": [Function],
          "getRelativeFormat": [Function],
        },
        "locale": "en",
        "messages": null,
        "now": [Function],
        "onError": [Function],
        "textComponent": "span",
        "timeZone": null,
      }
    }
    isSingleLine={false}
    onSelect={[Function]}
  >
    <ResponsiveElement
      onResize={[Function]}
      responsiveTo="window"
    >
      <div
        aria-label="Example of a Selectable Pill"
        className="pill-list"
        onBlur={[Function]}
        onKeyDown={[Function]}
        tabIndex="-1"
      >
        <VisuallyHiddenText
          id="terra-pill-visual-hidden-text"
          text="Terra.pills.pillListHint"
        >
          <span
            className="visually-hidden-text"
            id="terra-pill-visual-hidden-text"
          >
            Terra.pills.pillListHint
          </span>
        </VisuallyHiddenText>
        <InjectIntl(Pill)
          id="selectabla-pill"
          isBasicPill={false}
          key=".0"
          label="Selectable Pill"
          onRemove={[Function]}
          onSelect={[Function]}
        >
          <Pill
            id="selectabla-pill"
            intl={
              Object {
                "defaultFormats": Object {},
                "defaultLocale": "en",
                "formatDate": [Function],
                "formatHTMLMessage": [Function],
                "formatMessage": [Function],
                "formatNumber": [Function],
                "formatPlural": [Function],
                "formatRelative": [Function],
                "formatTime": [Function],
                "formats": Object {},
                "formatters": Object {
                  "getDateTimeFormat": [Function],
                  "getMessageFormat": [Function],
                  "getNumberFormat": [Function],
                  "getPluralFormat": [Function],
                  "getRelativeFormat": [Function],
                },
                "locale": "en",
                "messages": null,
                "now": [Function],
                "onError": [Function],
                "textComponent": "span",
                "timeZone": null,
              }
            }
            isBasicPill={false}
            label="Selectable Pill"
            onRemove={[Function]}
            onSelect={[Function]}
          >
            <ResponsiveElement
              onResize={[Function]}
              responsiveTo="window"
            >
              <div
                className="pill-container is-focusable is-selectable"
                data-terra-pill={true}
                data-terra-pills-show-focus-styles={true}
                id="selectabla-pill"
                onBlur={[Function]}
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                role="button"
                tabIndex="0"
              >
                <div
                  className="pill-label is-selectable"
                  onClick={[Function]}
                  title=""
                >
                  Selectable Pill
                </div>
                <VisuallyHiddenText
                  text="Terra.pills.pillHint.selectable"
                >
                  <span
                    className="visually-hidden-text"
                  >
                    Terra.pills.pillHint.selectable
                  </span>
                </VisuallyHiddenText>
              </div>
            </ResponsiveElement>
          </Pill>
        </InjectIntl(Pill)>
      </div>
    </ResponsiveElement>
  </SelectablePills>
</InjectIntl(SelectablePills)>
`;

exports[`Selectable Pill should render a selectable pill with roll up 1`] = `
<InjectIntl(SelectablePills)
  ariaLabel="Example of a Selectable Pill"
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": null,
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
  isSingleLine={true}
  onSelect={[Function]}
>
  <SelectablePills
    ariaLabel="Example of a Selectable Pill"
    intl={
      Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatHTMLMessage": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatPlural": [Function],
        "formatRelative": [Function],
        "formatTime": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralFormat": [Function],
          "getRelativeFormat": [Function],
        },
        "locale": "en",
        "messages": null,
        "now": [Function],
        "onError": [Function],
        "textComponent": "span",
        "timeZone": null,
      }
    }
    isSingleLine={true}
    onSelect={[Function]}
  >
    <ResponsiveElement
      onResize={[Function]}
      responsiveTo="window"
    >
      <div
        aria-label="Example of a Selectable Pill"
        className="pill-list"
        onBlur={[Function]}
        onKeyDown={[Function]}
        tabIndex="-1"
      >
        <VisuallyHiddenText
          id="terra-pill-visual-hidden-text"
          text="Terra.pills.pillListHint"
        >
          <span
            className="visually-hidden-text"
            id="terra-pill-visual-hidden-text"
          >
            Terra.pills.pillListHint
          </span>
        </VisuallyHiddenText>
        <InjectIntl(RollUpPill)
          isSingleLine={true}
          onSelectRollUp={[Function]}
          rollupCount={1}
        >
          <RollUpPill
            intl={
              Object {
                "defaultFormats": Object {},
                "defaultLocale": "en",
                "formatDate": [Function],
                "formatHTMLMessage": [Function],
                "formatMessage": [Function],
                "formatNumber": [Function],
                "formatPlural": [Function],
                "formatRelative": [Function],
                "formatTime": [Function],
                "formats": Object {},
                "formatters": Object {
                  "getDateTimeFormat": [Function],
                  "getMessageFormat": [Function],
                  "getNumberFormat": [Function],
                  "getPluralFormat": [Function],
                  "getRelativeFormat": [Function],
                },
                "locale": "en",
                "messages": null,
                "now": [Function],
                "onError": [Function],
                "textComponent": "span",
                "timeZone": null,
              }
            }
            isSingleLine={true}
            onSelectRollUp={[Function]}
            rollupCount={1}
          >
            <div
              className="roll-up-pill"
              data-terra-rollup-pill={true}
              data-terra-rollup-pill-show-focus-styles={true}
              id="rollup-pill"
              onClick={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="button"
              tabIndex="0"
            >
              Terra.pills.rollupPillLabel
            </div>
          </RollUpPill>
        </InjectIntl(RollUpPill)>
      </div>
    </ResponsiveElement>
  </SelectablePills>
</InjectIntl(SelectablePills)>
`;
